org.eclipse.m2e.jdt.tests/src/org/eclipse/m2e/jdt/tests/JpmsConfigurationTest.java

```diff
@@ -0,0 +1,163 @@
+  
+  @Test
+  public void testFileChangeUpdatesJPMSSettings() throws CoreException, IOException, InterruptedException {
+    ((MavenConfigurationImpl) MavenPlugin.getMavenConfiguration()).setAutomaticallyUpdateConfiguration(true);
+    setAutoBuilding(true);
+    
+    File pomFileFS = new File(FileLocator.toFileURL(getClass().getResource("/projects/compilerJpmsSettings/pom.xml")).getFile());
+    File argsSet1FS= new File(FileLocator.toFileURL(getClass().getResource("/projects/compilerJpmsSettings/argsSet1.xml")).getFile());
+    File argsSet2FS = new File(FileLocator.toFileURL(getClass().getResource("/projects/compilerJpmsSettings/argsSet2.xml")).getFile());
+    
+    IProject project = importProject(pomFileFS.getAbsolutePath());
+    waitForJobsToComplete();
+    
+    IJavaProject javaProject = (IJavaProject) project.getNature(JavaCore.NATURE_ID);
+    Map<String, String> jreAttributes = getJreContainerAttributes(javaProject);
+    Map<String, String> m2eAttributes = getM2eContainerAttributes(javaProject);
+    
+    assertEquals("11", javaProject.getOption(JavaCore.COMPILER_SOURCE, false));
+    
+    assertTrue(!jreAttributes.containsKey(ADD_EXPORTS_ATTR));
+    assertTrue(!jreAttributes.containsKey(ADD_OPENS_ATTR));
+    assertTrue(!jreAttributes.containsKey(ADD_READS_ATTR));
+    assertTrue(!jreAttributes.containsKey(PATCH_MODULE_ATTR));
+    
+    assertTrue(!m2eAttributes.containsKey(ADD_EXPORTS_ATTR));
+    assertTrue(!m2eAttributes.containsKey(ADD_OPENS_ATTR));
+    assertTrue(!m2eAttributes.containsKey(ADD_READS_ATTR));
+    assertTrue(!m2eAttributes.containsKey(PATCH_MODULE_ATTR));
+    
+    
+    
+    String contents = read(project, pomFileFS);
+    String argsSet1 = read(project, argsSet1FS);
+    String argsSet2 = read(project, argsSet2FS);
+    
+    IFile pomFileWS = project.getFile(pomFileFS.getName());
+    
+    String argsSet1Content = contents.replace(REPLACED_POM_STRING, argsSet1);
+    pomFileWS.setContents(new ByteArrayInputStream(argsSet1Content.getBytes()), true, false, null);
+    waitForJobsToComplete();
+    
+    jreAttributes = getJreContainerAttributes(javaProject);
+    m2eAttributes = getM2eContainerAttributes(javaProject);
+    
+    assertTrue(jreAttributes.get(ADD_EXPORTS_ATTR).equals(JRE_ADD_EXPORTS_VALUE1));
+    assertTrue(jreAttributes.get(ADD_OPENS_ATTR).equals(JRE_ADD_OPENS_VALUE1));
+    assertTrue(jreAttributes.get(ADD_READS_ATTR).equals(JRE_ADD_READS_VALUE1));
+    assertTrue(jreAttributes.get(PATCH_MODULE_ATTR).equals(JRE_PATCH_MODULE_VALUE1));
+    
+    assertTrue(m2eAttributes.get(ADD_EXPORTS_ATTR).equals(M2E_ADD_EXPORTS_VALUE1));
+    assertTrue(m2eAttributes.get(ADD_OPENS_ATTR).equals(M2E_ADD_OPENS_VALUE1));
+    assertTrue(m2eAttributes.get(ADD_READS_ATTR).equals(M2E_ADD_READS_VALUE1));
+    assertTrue(m2eAttributes.get(PATCH_MODULE_ATTR).equals(M2E_PATCH_MODULE_VALUE1));
+    
+    String argsSet2Content = contents.replace(REPLACED_POM_STRING, argsSet2);
```

Comment:
would be better to cover compilerArgument works as intended on the second test

Reply:
I do not handle compilerArgument anymore in the getCompilerAgument method. As the maven-compiler-plugin doc states, all the jpms arguments are in fact two arguments so no complete jpms argument can be provided using compilerArgument.

And i think it is not worth supporting the following cases:

```
<compilerArgument>--add-exports</compilerArgument>
<compilerArgs>
   <arg>somemodule/somepackage=someothermodule</arg>
</compilerArgs>

<compilerArgs>
   <arg>--add-exports</arg>
</compilerArgs>
<compilerArgument>somemodule/somepackage=someothermodule</compilerArgument>
```

Do you agree ?